src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -110,23 +150,232 @@ public override IReadOnlyList<TextChangeRange> GetChangeRanges(SourceText oldTex
             SourceText actualOldText;
-            if (_weakOldText.TryGetTarget(out actualOldText))
+            if (_info.WeakOldText.TryGetTarget(out actualOldText) && actualOldText == oldText)
+            {
+                // the supplied old text is the one we directly reference, so the changes must be the ones we have.
+                return _info.ChangeRanges;
+            }
+
+            // otherwise look to see if there are a series of changes from the old text to this text and merge them.
+            if (IsChangedFrom(oldText))
+            {
+                var changes = GetChangesBetween(oldText, this);
+                if (changes.Count > 1)
+                {
+                    return Merge(changes);
+                }
+            }
+
+            // the SourceText subtype for editor snapshots knows when two snapshots from the same buffer have the same contents
+            if (actualOldText != null && actualOldText.GetChangeRanges(oldText).Count == 0)
+            {
+                // the texts are different instances, but the contents are considered to be the same.
+                return _info.ChangeRanges;
+            }
+
+            return ImmutableArray.Create(new TextChangeRange(new TextSpan(0, oldText.Length), _newText.Length));
+        }
+
+        private bool IsChangedFrom(SourceText oldText)
+        {
+            for (var info = _info; info != null; info = info.Previous)
+            {
+                SourceText text;
+                if (info.WeakOldText.TryGetTarget(out text) && text == oldText)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static IReadOnlyList<ImmutableArray<TextChangeRange>> GetChangesBetween(SourceText oldText, ChangedText newText)
+        {
+            var list = new List<ImmutableArray<TextChangeRange>>();
+
+            var change = newText._info;
+            list.Add(change.ChangeRanges);
+
+            while (change != null)
             {
+                SourceText actualOldText;
+                change.WeakOldText.TryGetTarget(out actualOldText);
+
                 if (actualOldText == oldText)
                 {
-                    // same identity, so the changes must be the ones we have.
-                    return _changes;
+                    return list;
                 }
 
-                if (actualOldText.GetChangeRanges(oldText).Count == 0)
+                change = change.Previous;
+                if (change != null)
                 {
-                    // the bases are different instances, but the contents are considered to be the same.
-                    return _changes;
+                    list.Insert(0, change.ChangeRanges);
                 }
             }
 
-            return ImmutableArray.Create(new TextChangeRange(new TextSpan(0, oldText.Length), _newText.Length));
+            // did not find old text, so not connected?
```

Comment:
This can happen if the old text and new text are not actually related to each other through a chain on explicit WithChanges operations.